mono/mini/method-to-ir.c

```diff
@@ -7036,8 +7006,22 @@ is_supported_tail_call (MonoCompile *cfg, MonoMethod *method, MonoMethod *cmetho
 		supported_tail_call = FALSE;
 	if (cmethod->wrapper_type && cmethod->wrapper_type != MONO_WRAPPER_DYNAMIC_METHOD)
 		supported_tail_call = FALSE;
-	if (call_opcode != CEE_CALL)
+
+	switch (call_opcode) {
+	case CEE_CALL:
+		break;
+	case CEE_CALLI:
 		supported_tail_call = FALSE;
+		break;
+	case CEE_CALLVIRT:
+		if (!cfg->backend->have_op_tail_call_membase)
+			supported_tail_call = FALSE;
+		break;
+	default:
+		g_assertf (call_opcode == CEE_CALL || call_opcode == CEE_CALLVIRT || call_opcode == CEE_CALLI, "%s (%d)", mono_opcode_name (call_opcode), (int)call_opcode);
```

Comment:
That's funny, isn't this `g_assertf` always going to fail? Except if I am misreading something of course, but if it's CEE_CALL, CEE_CALLI or CEE_CALLVIRT then it's going to hit the other case, otherwise it asserts.

Reply:
Assert is nice because it provides builtin text.

Reply:
Asserts that come after a condition and always fail are not uncommon.